### PR TITLE
feature(backend): adding variables services for tests and transactions

### PR DESCRIPTION
**Pull request by xoscar**
This PR adds the variables services to both transactions and tests to return the environment variables required from the test, what is already set and which ones are missing.

## Changes

- Adds new API services for both transactions and tests
- Creates a new Variables struct to handle environment and spec env interpolations

## Fixes

- #1643 

## Checklist

- [x] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

# Diagrams
The following sequence diagrams describe how the two new services will work to get the missing variables from a test or a transaction.

## Test:

```mermaid
sequenceDiagram
    Frontend->>+Server: GET /${testId}/variables?environmentId=<envId>
    Server->>+Database: Fetch Test
    Database-->>-Server: Return test
    alt is environment Id present
        Server->>+Database: Fetch Environment
        Database-->>-Server: Return Environment
        Server->>Server: Initializes env from database result
    else
        Server->>Server: Initializes empty env
    end
    Server->>Server: Adds test outputs to the env
    Server->>Server: Analyzes test looking for interpolations
    Server->>Server: Generates variables based on interpolations
    Server->>Server: Matches variables with env to find missing entries
    Server-->>-Frontend: Returns env, variables, and missing
```

## Transaction:

```mermaid
sequenceDiagram
    Frontend->>+Server: GET /{tId}/variables?environmentId=<envId>
    Server->>+Database: Fetches Transaction
    Database-->>-Server: Returns Transaction
    alt is environment Id present
        Server->>+Database: Fetch Environment
        Database-->>-Server: Return Environment
        Server->>Server: Initializes env from database result
    else
        Server->>Server: Initializes empty env
    end
    loop For Every Test in Transaction
        Server->>+Database: Fetches Test
        Database-->>-Server: Returns Test
        Server->>Server: Adds test outputs to the env
        Server->>Server: Analyzes test looking for interpolations
        Server->>Server: Generates variables based on interpolations
        Server->>Server: Matches variables with env to find missing entries
        Server->>Server: Update environment store for next tests
    end
    Server-->>-Frontend: Returns env, variables and missing
```
https://www.loom.com/share/6eb73136470947d8ad1966629392bea9

### CODE REVIEW

File: server/http/controller.go [hunk 1]
```diff
@@ -1221,3 +1225,92 @@ func (c *controller) TestConnection(ctx context.Context, dataStore openapi.DataS
...
+	if err != nil {
+		return openapi.Response(http.StatusInternalServerError, err.Error()), err
+	}
+
+	return openapi.Response(200, c.mappers.Out.TestVariables(testVariables)), nil
+}
+
+func processTestVariables(ctx context.Context, test model.Test, variables expression.Variables, triggerRegistry *trigger.Registry) (expression.TestVariables, error) {
+	environmentVariables := variables.GetEnvironmentVariables(test)
+	specVariables, err := variables.GetSpecsVariables(test)
```
Comment (jorgeepc) [inline server/http/controller.go:1261]: Do we need a similar step for `Outputs`? Similar to specs we can reference a `env:` var in an output value, right?

Comment (xoscar) [reply server/http/controller.go:1261]: You are right, I'm adding it

**Review by jorgeepc — COMMENTED** [review]
Also saw in you loom that we get duplicated variables in the response. What do you think about getting unique values?

File: server/http/controller.go (hunk 1 above)
Comment (xoscar) [reply server/http/controller.go:1261]: Done 😄

**Comment by kdhamric** [comment]
One requirement is to provide a default for each missing variable based on the previous run of this same test - I did not see that covered in the loom video:

requirement:
for both UI and CLI, the promoting screen should show the variable name and ask for the value. If there have been previous runs of the test, the value of the variable for the last run should be used if available.

**Comment by xoscar** [comment]
> One requirement is to provide a default for each missing variable based on the previous run of this same test - I did not see that covered in the loom video:
> 
> requirement: for both UI and CLI, the promoting screen should show the variable name and ask for the value. If there have been previous runs of the test, the value of the variable for the last run should be used if available.

@kdhamric this is only the API implementation to grab the missing variables, we can ensure that connection when working on the client side. This is not a fully ready feature yet.

**Comment by xoscar** [comment]
> Also saw in you loom that we get duplicated variables in the response. What do you think about getting unique values?

Done 😄

**Review by danielbdias — COMMENTED** [review]
At first glance, the code is ok for me! I have two questions about it: 
1) the idea of using [samber/lo](https://github.com/samber/lo) is because of some parts of repetitive code, like in the grpc and HTTP executor, right? 
2) Do you have an example of an API call that we could do to test this feature? (rationale behind it: if we can make a curl exercising this use case, maybe we can add a dogfood/tracetesting case in our CI later)

File: server/executor/trigger/grpc.go [hunk 2]
```diff
@@ -144,6 +145,52 @@ func (t *grpcTriggerer) Resolve(ctx context.Context, test model.Test, opts *Trig
...
+	grpc := test.ServiceUnderTest.GRPC
+
+	if grpc == nil {
+		return triggerVariables, fmt.Errorf("no settings provided for GRPC triggerer")
+	}
+	address, err := executor.StatementTermsByType(WrapInQuotes(grpc.Address, "\""), expression.EnvironmentType)
+	if err != nil {
+		return triggerVariables, err
+	}
+	triggerVariables = triggerVariables.MergeStringArray(address)
```
Comment (danielbdias) [inline server/executor/trigger/grpc.go:159]: Small / not important suggestion: to make it easier to read it makes sense to group the guard clauses if the statements that we are checking, like the `grpc := ...` line and `if grpc == nil { ...` line

```suggestion
	triggerVariables := expression.VariablesMap{}
	
	grpc := test.ServiceUnderTest.GRPC
	if grpc == nil {
		return triggerVariables, fmt.Errorf("no settings provided for GRPC triggerer")
	}
	
	address, err := executor.StatementTermsByType(WrapInQuotes(grpc.Address, "\""), expression.EnvironmentType)
	if err != nil {
		return triggerVariables, err
	}
	
	triggerVariables = triggerVariables.MergeStringArray(address)
```

File: server/executor/trigger/http.go [hunk 3]
```diff
@@ -199,3 +199,70 @@ func mapResp(resp *http.Response) model.HTTPResponse {
...
+	if http == nil {
+		return triggerVariables, fmt.Errorf("no settings provided for HTTP triggerer")
+	}
+
+	urlVariables, err := executor.StatementTermsByType(WrapInQuotes(http.URL, "\""), expression.EnvironmentType)
+
+	if err != nil {
+		return triggerVariables, err
+	}
+	triggerVariables = triggerVariables.MergeStringArray(urlVariables)
```
Comment (danielbdias) [inline server/executor/trigger/http.go:216]: Same suggestion as the gRPC endpoint:

```suggestion
	triggerVariables := expression.VariablesMap{}
	
	http := test.ServiceUnderTest.HTTP
	if http == nil {
		return triggerVariables, fmt.Errorf("no settings provided for HTTP triggerer")
	}

	urlVariables, err := executor.StatementTermsByType(WrapInQuotes(http.URL, "\""), expression.EnvironmentType)
	if err != nil {
		return triggerVariables, err
	}
	
	triggerVariables = triggerVariables.MergeStringArray(urlVariables)
```

File: server/executor/trigger/grpc.go (hunk 2 above)
Comment (xoscar) [reply server/executor/trigger/grpc.go]: Done 😄

**Comment by xoscar** [comment]
Hey @danielbdias thanks for the review man, the PR should be ready for its final review now 😄

**Review by danielbdias — APPROVED** [review]
:shipit:

**Pull request merged**